Load test horizontal pod autoscaler (#2009)
kyma-project/kyma: tools/load-test/README.md

# Load Test

## Overview

> The purpose of the load test is to verify the right execution of the Horizonatl Pod Autoscaling in **Kyma**. It runs a **golang** app which stress a function by making thousands of http post request in set period of time defined in an environment variable. At the end of the execution of the test a notification is send to a Slack channel with the final output of the load test.

Reviewer: There is no info that tool is used for Kubeless. I understood it only by assuming that `function` word used here to refer to Function CR. And also I saw some Kubeless env variables

Reviewer 2: I included a reference to Kubeless here https://github.com/kyma-project/kyma/pull/2009/commits/8d218456fb22fc3c7c3eab4d58c66969a4be0d88. My belief is this tool can be extended for more areas(rather than just functions) to load test Kyma in future.

# Load Test

## Overview

The purpose of the load test is to verify the execution of the Horizontal Pod Autoscaling of functions deployed with the help of Kubeless in **Kyma**. It runs a **golang** app which stress a function by making thousands of HTTP POST request in set period of time defined in an environment variable. At the end of the execution of the test a notification is send to a Slack channel with the final output of the load test.

## Installation

```bash
echo "Installing helm chart..."
helm install --set slackEndpoint="${SLACK_ENDPOINT}" \
             --set slackClientToken="${SLACK_CLIENT_TOKEN}" \
             --set slackClientChannelId="${SLACK_CLIENT_CHANNEL_ID}" \
             --set loadTestExecutionTimeout="$LT_TIMEOUT" \
             ${LT_FOLDER_CHART} \
             --namespace=kyma-system \
             --name=load-test
```

The configuration needed to execute the **load test** is as follows:

 | Name | Value | Description |
 |------|---------------|-------------|
**slackEndpoint** |`-`| A webhook slack url.
**slackClientToken** |`-`|  A token which will be part of the **slackEndpoint**.
**slackClientChannelId** |`#channelId`| ID of the Slack channel.

> **loadTestExecutionTimeout** |`30`| Time to finish the test otherwise it will timeout its execution. All the related metrics to be sent to the Slack channel are collected after the timeout.

Reviewer 3: ```suggestion
**loadTestExecutionTimeout** |`30`| Time after which the test execution timeout occurs. All the related metrics to be sent to the Slack channel are collected after the timeout.
```

### Environment Variables

> To run the **load test** either in a cluster or a local Minikube you will need to set the above parameters for the installation of the helm chart. However in a cluster it is advisable to have environment variables as it is shown above.

Reviewer 3: ```suggestion
To run the load test either on a cluster or on a local Minikube, you must set the Helm chart installation parameters. However, on a cluster it is advisable to have environment variables set to default values.
```

Reviewer 3: This section does not fit here. Please move it above the table with the environment variables.

> ```bash

Reviewer 3: What is that for?

### Test config
# export SLACK_CLIENT_TOKEN='replace by the slack token'
# export SLACK_CLIENT_CHANNEL_ID='replace by the slack #channel'
# export SLACK_ENDPOINT='replace by the webhook http endpoint https://endpoint_here'
# export LT_FOLDER_CHART='HOME_KYMA_PROJECT/kyma/tools/load-test/deploy/chart/load-test'
# export LT_TIMEOUT=30
```

## Development

> 

Reviewer 3: Please add an introductory sentence explaining what this section presents. I'm not really sure what the bullet list below describes.

- **load-test/k8syaml**  contains all the Kubernetes resources needed to deploy the function.

> - **main.go** All the logic of the **load test** can be found in this file. It can be built as follows:

Reviewer 3: ```suggestion
- **main.go** contains all the logic of the load test. You can build it with these commands:
```

 
 `GOOS=linux GOARCH=amd64 go build -o ./bin/app`(Mac)
 `CGO_ENABLED=0 go build -o ./bin/app`(Linux)
 

> - **load-test/Dockerfile** Needed to build the docker image. The image can be built as a follows:

Reviewer 3: ```suggestion
- **load-test/Dockerfile** is a file needed to build the Docker image. To build the image, run this command:
```

`docker build -t load-test .`

> - **load-test/deploy/chart** contains the the chart which installs the test code which in turn stresses the function.

Reviewer 3: ```suggestion
- **load-test/deploy/chart** contains the chart that installs the test code which stresses the function.
```

# Load Test

## Overview

The load test verifies the execution of the Horizontal Pod Autoscaling of functions deployed using Kubeless in Kyma. It runs a Golang application which stresses a function by making thousands of HTTP POST requests in a set period of time defined in an environment variable. At the end of the test execution, a notification with the output is sent to a Slack channel.

## Installation

> As the load test has been implemented as `helm chart` the script below is the way how the test must be installed.

Reviewer 3: ```suggestion
As the load test is implemented as a Helm chart, run the following script to install the test:
```

```bash
echo "Installing helm chart..."
helm install --set slackEndpoint="${SLACK_ENDPOINT}" \
             --set slackClientToken="${SLACK_CLIENT_TOKEN}" \
             --set slackClientChannelId="${SLACK_CLIENT_CHANNEL_ID}" \
             --set loadTestExecutionTimeout="$LT_TIMEOUT" \
             ${LT_FOLDER_CHART} \
             --namespace=kyma-system \
             --name=load-test
```

These are the parameters you must configure to execute the load test:

 | Parameter | Value | Description |
 |------|---------------|-------------|
**slackEndpoint** |`-`| A webhook Slack URL.
**slackClientToken** |`-`|  A token which is a part of the **slackEndpoint**.
**slackClientChannelId** |`#channelId`| ID of the Slack channel.
**loadTestExecutionTimeout** |`30`| Time after which the test execution timeout occurs. All the related metrics to be sent to the Slack channel are collected after the timeout.

> ### Environment Variables

Reviewer 3: I'd get rid of this heading.

> To run the load test either on a cluster or on a local Minikube, you must set the Helm chart installation parameters. However, on a cluster it is advisable to have environment variables set to default values.

Reviewer 3: I'd move this section above the table with parameters (and also above the "These are the parameters..." sentence).

> The `bash` instructions below might be executed in case it is wanted to run the test either locally on miniukube or in a cluster.

Reviewer 3: ```suggestion
You can also define the values of these parameters by running this command:
```

```bash
### Test config
# export SLACK_CLIENT_TOKEN='replace by the slack token'
# export SLACK_CLIENT_CHANNEL_ID='replace by the slack #channel'
# export SLACK_ENDPOINT='replace by the webhook http endpoint https://endpoint_here'
# export LT_FOLDER_CHART='HOME_KYMA_PROJECT/kyma/tools/load-test/deploy/chart/load-test'
# export LT_TIMEOUT=30
```

## Development

> The main components of the load test can be divided into four and developers might be interested in adjust some of them to their requirements, such as the function definition or even the logic of the test itself throught the `.go` file.

Reviewer 3: ```suggestion
These are the main components of the load test:
```

- **load-test/k8syaml**  contains all the Kubernetes resources needed to deploy the function.

- **load-test.go** contains all the logic of the load test. You can build it with these commands:
 

>  `GOOS=linux GOARCH=amd64 go build -o ./bin/app`(Mac)

Reviewer 3: `
```
GOOS=linux GOARCH=amd64 go build -o ./bin/app (Mac)
```
`

# Load Test

## Overview

The load test verifies the execution of the Horizontal Pod Autoscaling of functions deployed using Kubeless in Kyma. It runs a Golang application which stresses a function by making thousands of HTTP POST requests in a set period of time defined in an environment variable. At the end of the test execution, a notification with the output is sent to a Slack channel.

## Installation

> As the load test is implemented as a Helm chart, run the following script to install the test:

Reviewer 3: ```suggestion
Run the following script to install the test:
```

```bash
echo "Installing helm chart..."
helm install --set slackEndpoint="${SLACK_ENDPOINT}" \
             --set slackClientToken="${SLACK_CLIENT_TOKEN}" \
             --set slackClientChannelId="${SLACK_CLIENT_CHANNEL_ID}" \
             --set loadTestExecutionTimeout="$LT_TIMEOUT" \
             ${LT_FOLDER_CHART} \
             --namespace=kyma-system \
             --name=load-test
```

To run the load test either on a cluster or on a local Minikube, you must set the Helm chart installation parameters. However, on a cluster it is advisable to have environment variables set to default values.

These are the parameters you must configure to execute the load test:

 | Parameter | Value | Description |
 |------|---------------|-------------|
**slackEndpoint** |`-`| A webhook Slack URL.
**slackClientToken** |`-`|  A token which is a part of the **slackEndpoint**.
**slackClientChannelId** |`#channelId`| ID of the Slack channel.
**loadTestExecutionTimeout** |`30`| Time after which the test execution timeout occurs. All the related metrics to be sent to the Slack channel are collected after the timeout.

> You can also define the values of these parameters by running this command:

Reviewer 3: ```suggestion
You can also define the values of these parameters while running this command:
```
